Realm/RLMObjectBase.mm

```diff
@@ -93,7 +93,8 @@ - (void)dealloc {
 }
 
 static id coerceToObjectType(id obj, Class cls, RLMSchema *schema) {
-    if ([obj isKindOfClass:cls]) {
+    if (([obj isKindOfClass:cls] && ![(id)cls isEmbedded]) ||
+        ([(id)cls isEmbedded] && [obj respondsToSelector:@selector(realm)] && ![obj realm])) {
```

Comment:
```suggestion
    if ([obj isKindOfClass:cls] && (![(id)cls isEmbedded] || ![obj realm])) {
```

I think what you have would call `isEmbedded` on classes that don't implement it and would accept objects of any type as long as they're an unmanaged embedded object. This also points at some missing tests for initializing an unmanaged object with a value that has an object of a different type in an embedded object field (which copies that object into an unmanaged object of the correct type if all of the required fields are present and have the correct values, and throws an exception if it can't).